monai/networks/utils.py

```diff
@@ -837,3 +838,67 @@ def replace_modules_temp(
         # revert
         for name, module in replaced:
             _replace_modules(parent, name, module, [], strict_match=True, match_device=match_device)
+
+
+class AutoDeviceAdapt:
+    """
+    In case of out-of-memory error when running a callable module ``mod``,
+    this class will take one of the following actions:
+
+        - change the "cpu_thresh" if ``mod`` supports this option.
+        - change ``mod.device`` to "cpu" temporarily if ``mod.device`` is set to None.
+        - move ``inputs`` to "cpu" if ``inputs`` is a torch tensor.
+        - raise an OOM error.
+
+    The class is designed to make the pipeline robust to outliers of unusually large ``inputs`` when
+    calling ``mod(inputs, ...)`` with heterogeneous ``inputs`` on CUDA.
+    Since the error handling is expensive, if the OOM happens frequently, ``mod`` should be adjusted instead of
+    relying on this class.
+    """
+
+    def __init__(self, mod):
+        """
+        Args:
+            mod: the callable module to be adjusted. If it's first input argument is a torch Tensor,
+                this class may try to move it to "cpu" if it's not already on "cpu".
+        """
+        self.mod = mod
+        self.has_device = hasattr(self.mod, "device")
+        self.has_cpu_thresh = hasattr(self.mod, "cpu_tresh")
+
+    def __call__(self, inputs, *args, **kwargs):
+        try:
+            return self.mod(inputs, *args, **kwargs)
+        except RuntimeError as e:
+            if "OutOfMemoryError" not in str(type(e).__name__) or "memory" not in f"{e}".lower():
+                raise e  # this class only tries to handle OOM
+            if isinstance(inputs, torch.Tensor) and not inputs.is_cuda:
+                raise e  # inputs is not currently using cuda, no auto-tune
+            if self.has_device and self.mod.device is not None:
+                raise e  # user explicitly set device="cuda"/"cpu"/..., no auto-tune
+            if self.has_cpu_thresh:
+                if isinstance(inputs, torch.Tensor) and self.mod.cpu_thresh <= inputs.shape[2:].numel():
+                    warnings.warn(
+                        f"self.mod.cpu_thresh is set to {self.mod.cpu_thresh}, "
+                        f" and input size: {inputs.shape[2:].numel()}"
+                    )
+                raise e  # self.mod currently runs on cpu according to self.mod.cpu_thresh
+            warnings.warn(
+                f"{type(self.mod)} failed {e} device={self.mod.device}, "
+                f" attempting CPU, image dim {inputs.shape}. The error handling is expensive, "
+                f" please review {type(self.mod)} settings if this message appears frequently."
+            )
+            if not self.has_cpu_thresh:  # mod doesn't support dynamic input size based device allocation
+                if isinstance(inputs, torch.Tensor):
+                    inputs = inputs.to("cpu")
```

Comment:
I don't think we should do it,  if "model" weights are on GPU, it won't accept inputs on CPU, and will crash, so then we need to worry where the model weights are.